# 01 - Data Foundation

## 1. Objectives

The objectives of this notebook are to:

- Load and inspect the raw dataset without modifying the source file.
- Validate the dataset structure, dimensions, and data types.
- Assess missing values, duplicates, and unexpected values.
- Validate the uniqueness of the lead identifier.
- Understand the distribution and meaning of `CallStatus`.
- Create consistent analytical fields for lead quality and time-based analysis.
- Document data limitations that may affect subsequent conclusions.
- Produce a clean analytical dataset for use in the remaining notebooks.

### Expected outcome

At the end of this notebook, we should have a clearly documented and analysis-ready dataset, along with an explicit record of any data-quality limitations that need to be considered in later analysis.

## 2. Importing Libraries

The analysis uses Pandas and NumPy for data preparation, Matplotlib and Seaborn for visualization, and SciPy for statistical testing.

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from scipy import stats

sns.set_theme(style="whitegrid")
pd.set_option("display.max_columns", None)
pd.set_option("display.float_format", "{:,.2f}".format)

### 3. Load the Raw Dataset

The source file is loaded without modification. All subsequent transformations will be performed on a working dataframe so that the original dataset remains unchanged.

In [2]:
df = pd.read_excel("../data/raw/Analyst_case_study_dataset_1_.xls")
df.head()

,LeadCreated,FirstName,Email,VendorLeadID,CallStatus,WidgetName,PublisherZoneName,PublisherCampaignName,AddressScore,PhoneScore,AdvertiserCampaignName,State,DebtLevel,IP Address,Partner,ReferralDomain,MarketingCampaign,AdGroup,Keyword,SearchQuery,ReferralURL,ReferralURL Parameters,LandingPageURL,Landing Page URL Parameters
0,2009-07-01 01:24:29.563,Dorinda,kanani@sandwichisles.net,FDF81FDA-A649-437B-B99C-FDDE74F7971B,NaN,w-302252-DebtReduction1-1DC-CreditSolutions,TopLeft-302252,DebtReductionInc,NaN,5.00,creditsolutions-branded-shortform,HI,30001-50000,NaN,google,www.google.com,Debt Holding Tank,Holding Tank - Debt,Debt specialists,debt specialists,http://www.google.com/search,sourceid=navclient&aq=1&oq=debt sp&ie=UTF-8&rl...,http://www.debtreductioninc.com/index8.html,utm_source=google&utm_medium=CPC&utm_content=H...
1,2009-04-13 15:27:35.990,Presetta,clerk2@ustconline.net,4190ACB7-5026-416C-B987-ED8AD427D5E6,NaN,w-300250-DebtReduction1-1DC-Head2,TopLeft-302252,DebtReductionInc,NaN,NaN,Debt Settlement1 Master,AL,90000-100000,NaN,AdKnowledge,NaN,Financial Services,Consolidate,NaN,NaN,NaN,NaN,http://www.debtreductioninc.com/index8.html,utm_source=AdKnowledge&utm_medium=CPC&utm_cont...
2,2009-04-21 07:22:59.000,Gina,wagoner_gina@yahoo.com,hFg80jf_R0CRN55hdhWILw,Unable to contact - Bad Contact Information,w-300250-DebtReduction1-1DC-Head2,TopLeft-302252,DebtReductionInc,NaN,NaN,Debt Settlement1 Master,NY,20001-30000,NaN,AdKnowledge,us.mc582.mail.yahoo.com,Financial Services,Consolidate,NaN,NaN,http://us.mc582.mail.yahoo.com/mc/showMessage,&fid=Inbox&sort=date&order=down&startMid=0&.ra...,http://www.debtreductioninc.com/index8.html,utm_source=AdKnowledge&utm_medium=CPC&utm_cont...
3,2009-08-03 19:39:02.000,Kari,usa4ley@yahoo.com,jB01QgYZxkWArI9jWxuufw,Contacted - Doesn't Qualify,w-302252-DebtReduction1-1DC-white,TopLeft-302252,DebtReductionInc,5.00,3.00,Debt Settlement1 Master,WA,20001-30000,NaN,Google,norwich.kijiji.com,DebtReductionInc,Lower Payments,NaN,NaN,http://norwich.kijiji.com/c-Cars-vehicles-Cars...,NaN,http://www.debtreductioninc.com/index12.html,utm_source=Google&utm_medium=cpc&utm_campaign=...
4,2009-04-13 10:18:37.490,Stephanie,sr1lambert@embarqmail.com,D5B32074-458E-40EC-B185-1FEF20AC626D,NaN,w-300250-DebtReduction1-1DC-BlueMeter,TopLeft-302252,DebtReductionInc,NaN,NaN,Debt Settlement1 Master,TX,More_than_100000,NaN,Google,NaN,DebtReductionInc,Debt Reduction,NaN,NaN,NaN,NaN,http://www.debtreductioninc.com/index8.html,utm_source=Google&utm_medium=cpc&utm_campaign=...


In [3]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 3021 entries, 0 to 3020
Data columns (total 24 columns):
 #   Column                       Non-Null Count  Dtype         
---  ------                       --------------  -----         
 0   LeadCreated                  3021 non-null   datetime64[ns]
 1   FirstName                    3021 non-null   object        
 2   Email                        3021 non-null   object        
 3   VendorLeadID                 3013 non-null   object        
 4   CallStatus                   881 non-null    object        
 5   WidgetName                   3021 non-null   object        
 6   PublisherZoneName            3021 non-null   object        
 7   PublisherCampaignName        3021 non-null   object        
 8   AddressScore                 1171 non-null   float64       
 9   PhoneScore                   1393 non-null   float64       
 10  AdvertiserCampaignName       3021 non-null   object        
 11  State                        3021 non-null 

In [4]:
df.describe()

,LeadCreated,AddressScore,PhoneScore,IP Address
count,3021,"1,171.00","1,393.00",0.00
mean,2009-06-25 01:25:50.416092416,4.29,3.90,NaN
min,2009-04-02 18:52:41,1.00,1.00,NaN
25%,2009-05-13 22:56:40.676999936,4.00,3.00,NaN
50%,2009-06-18 19:45:45,5.00,4.00,NaN
75%,2009-08-03 12:21:34.972999936,5.00,5.00,NaN
max,2009-09-30 21:00:20.080000,5.00,5.00,NaN
std,NaN,1.35,1.10,NaN


In [5]:
df.columns.tolist()

['LeadCreated',
 'FirstName',
 'Email',
 'VendorLeadID',
 'CallStatus',
 'WidgetName',
 'PublisherZoneName',
 'PublisherCampaignName',
 'AddressScore',
 'PhoneScore',
 'AdvertiserCampaignName',
 'State',
 'DebtLevel',
 'IP Address',
 'Partner',
 'ReferralDomain',
 'MarketingCampaign',
 'AdGroup',
 'Keyword',
 'SearchQuery',
 'ReferralURL',
 'ReferralURL Parameters',
 'LandingPageURL',
 'Landing Page URL Parameters']

In [6]:
df.shape

(3021, 24)

## 4. Data Quality Audit

Before analysing lead quality, the dataset needs to be checked for issues that could affect the results.

The audit focuses on missing values, duplicate records, identifier uniqueness, and the values present in key categorical fields. These checks help distinguish genuine characteristics of the data from potential data-quality problems.

### Missing Values

In [7]:
missing = (df.isna().sum().to_frame("missing_count"))
missing["missing_pct"] = (missing["missing_count"] / len(df) * 100)
missing = missing.sort_values("missing_count", ascending=False)

missing

,missing_count,missing_pct
IP Address,3021,100.00
CallStatus,2140,70.84
Keyword,2042,67.59
AddressScore,1850,61.24
SearchQuery,1756,58.13
PhoneScore,1628,53.89
ReferralURL Parameters,738,24.43
ReferralDomain,515,17.05
ReferralURL,515,17.05
AdGroup,272,9.00


In [8]:
df[["CallStatus", "AddressScore", "PhoneScore"]].isna().sum()

CallStatus      2140
AddressScore    1850
PhoneScore      1628
dtype: int64

### Duplicate Records

In [9]:
duplicate_rows = df.duplicated().sum()
duplicate_rows

0

### Lead Identifier Uniqueness

In [10]:
print(f"Total rows: {len(df):,}")
print(f"Unique VendorLeadID values: {df['VendorLeadID'].nunique():,}")
print(f"Missing VendorLeadID values: {df['VendorLeadID'].isna().sum():,}")

Total rows: 3,021
Unique VendorLeadID values: 3,012
Missing VendorLeadID values: 8


In [11]:
duplicate_leads = df["VendorLeadID"].duplicated().sum()
print(f"Duplicate VendorLeadID values: {duplicate_leads:,}")

Duplicate VendorLeadID values: 8


### Key Categorical Value Audit

In [12]:
categorical_columns = ["CallStatus", "WidgetName", "PublisherZoneName", "PublisherCampaignName", "AdvertiserCampaignName"]

for column in categorical_columns:
    print(f"\n{column}")
    print("-" * len(column))
    print(df[column].value_counts(dropna=False))


CallStatus
----------
CallStatus
NaN                                            2140
Closed                                          245
Contacted - Doesn't Qualify                     212
Unable to contact - Bad Contact Information     191
EP Confirmed                                    130
Contacted - Invalid Profile                      85
EP Sent                                          16
EP Received                                       2
Name: count, dtype: int64

WidgetName
----------
WidgetName
w-302252-DebtReduction1-1DC-CreditSolutions     1054
w-302252-DebtReduction1-1DC-white                430
w-300250-DebtReduction1-1DC                      348
w-302252-DebtReduction1-1DC                      272
w-302252-DebtReduction1-1DC-yellowarrow-blue     232
w-302252-DebtReduction1-1DC-yellowarrow-dark     135
w-300250-DebtReduction1-1DC-BlueMeter             92
w-300250-DebtReduction1-1DC-Head2                 89
w-300250-DebtReduction1-2DC-BlueMeter             87
w-300250-Debt

### Lead Outcome Distribution

In [13]:
call_status_counts = (df["CallStatus"].value_counts(dropna=False).rename_axis("CallStatus").reset_index(name="LeadCount"))
call_status_counts["LeadShare"] = (call_status_counts["LeadCount"] / len(df) * 100)
call_status_counts

,CallStatus,LeadCount,LeadShare
0,NaN,2140,70.84
1,Closed,245,8.11
2,Contacted - Doesn't Qualify,212,7.02
3,Unable to contact - Bad Contact Information,191,6.32
4,EP Confirmed,130,4.30
5,Contacted - Invalid Profile,85,2.81
6,EP Sent,16,0.53
7,EP Received,2,0.07


### Date Coverage

In [14]:
print(df["LeadCreated"].min())
print(df["LeadCreated"].max())
print(df["LeadCreated"].dtype)

2009-04-02 18:52:41
2009-09-30 21:00:20.080000
datetime64[ns]


In [15]:
df[["AddressScore", "PhoneScore", "DebtLevel"]].describe()

,AddressScore,PhoneScore
count,"1,171.00","1,393.00"
mean,4.29,3.90
std,1.35,1.10
min,1.00,1.00
25%,4.00,3.00
50%,5.00,4.00
75%,5.00,5.00
max,5.00,5.00


In [16]:
print("AddressScore values:")
print(sorted(df["AddressScore"].dropna().unique()))

print("\nPhoneScore values:")
print(sorted(df["PhoneScore"].dropna().unique()))

AddressScore values:
[1.0, 2.0, 3.0, 4.0, 5.0]

PhoneScore values:
[1.0, 2.0, 3.0, 4.0, 5.0]


## Final Conclusion

The initial data audit establishes the structure, completeness, and reliability of the dataset before performing lead quality analysis.

The dataset contains **3,021 lead records across 24 columns**, covering the period from **April 2, 2009 to September 30, 2009**. The data contains a mixture of lead attributes, acquisition information, campaign details, geographic information, and outcome-related fields.

### Data Quality Findings

`VendorLeadID` is the intended lead identifier. There are **8 missing identifier values**, while the remaining non-null identifiers are not completely unique, with **3,012 unique non-null IDs across 3,021 records**. This should be considered when using the identifier for record-level analysis.

No completely duplicated rows were identified in the dataset, reducing the likelihood that exact duplicate records will materially distort aggregate lead-level metrics.

Several fields contain substantial missingness. The most significant examples are:

- `IP Address`: **100% missing**
- `CallStatus`: **70.84% missing**
- `Keyword`: **67.59% missing**
- `AddressScore`: **61.24% missing**
- `Referring Keyword String`: **58.13% missing**
- `PhoneScore`: **53.89% missing**

The missing `CallStatus` values are particularly important because `CallStatus` contains the advertiser's lead outcome information and is used to identify `Closed` leads. These records should therefore not be automatically classified as either successful or unsuccessful outcomes without additional evidence.

The `IP Address` field is completely missing and therefore does not provide usable analytical information in its current form.

### Dataset Structure

The categorical audit shows that the dataset contains:

- **14 WidgetName variants**
- **2 PublisherZoneName values**
- **2 PublisherCampaignName values**
- **2 AdvertiserCampaignName values**

The limited number of publisher and advertiser campaign variants makes these variables suitable for direct segment comparisons in the subsequent analysis.

### Lead Outcome Structure

`CallStatus` contains both positive and negative outcome states as well as intermediate progression states. There are **245 Closed leads**, while a large proportion of records have no recorded `CallStatus`.

This means that the dataset provides a usable basis for analyzing observed Closed outcomes, but the missing outcome information must remain an explicit limitation throughout the analysis.

### Overall Conclusion

The dataset is sufficiently structured to proceed with lead quality analysis, but several data-quality limitations need to be carried forward. In particular, missing outcome information, incomplete lead identifiers, and missing scoring/acquisition fields can affect the interpretation of segment-level results.

The most important analytical principle established from this audit is that missing values should be preserved and interpreted according to their business meaning rather than being automatically converted into positive or negative outcomes.

With the data structure and limitations established, the next analysis focuses on the advertiser's primary success measure, **Closed Rate**, and examines how it changes over time and across relevant lead and acquisition segments.